### ISSUE TITLE
Understanding 2.5.8 Target Size (minimum) can generate false positives

### ISSUE THREAD

**Issue by dd8** [w3c/wcag#3615]
The current wording of the 2.5.8 Target Size (minimum) understanding document can generate false positives with SVGs:
https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html

The following SVG graphic has a 10px wide clickable blue border, a large non-clickable area corresponding to fill=none, then an small clickable 20px x 20px red box in the middle of the non-clickable area with a 220px non-clickable gap on all sides.

```
<svg width="500" height="500" viewBox="0 0 500 500"> 
<a href="one"> <rect stroke-width="10px" stroke="blue" fill="none" x="10" y="0" width="480" height="480"></rect> </a> 
<a href="two"> <rect fill="red" x="240" y="240" width="20" height="20"></rect> </a> </svg>
```

Hit testing in HTML works differently to hit testing in SVG - in SVG it defaults to pointer-events: visiblePainted which is not how HTML behaves:
https://svgwg.org/svg2-draft/interact.html#PointerEventsProp

This is further complicated by the fact there's no definition of how hit testing actually works in HTML - this is an open issue in CSS:
https://drafts.csswg.org/css-ui-4/#pointer-events-control
https://github.com/w3c/csswg-drafts/issues/2325
https://github.com/w3c/csswg-drafts/issues/4326

**Comment by patrickhlauke** [comment]
the point of the SC is to avoid accidentally hitting the wrong/adjacent target. the target itself - and how easy/hard it is to activate - is in fact irrelevant. so arguably that SVG passes the requirement just fine - even if you had another target butting up right against the border, it would count as having sufficient spacing.

and yes, it's certainly far from real-world ideal...

**Comment by dd8** [comment]
Here's why I think it fails the current wording:

> "For a target to be "at least 24 by 24 CSS pixels", it must be possible to draw a solid 24 by 24 CSS pixel square, aligned to the horizontal and vertical axis such that the square is completely within the target"

- the red target area is a 20x20 pixel square so doesn't meet the size requirement 
- the blue target area is only 10 px wide so also doesn't meet the size requirement

> "Spacing: Undersized targets (those less than 24 by 24 CSS pixels) are positioned so that if a 24 CSS pixel diameter circle is centered on the [bounding box](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html#dfn-minimum-bounding-box) of each, the circles do not intersect another target or the circle for another undersized target"

The circles of both targets do intersect because they're centred on the bounding box of both targets (and the centres of both bounding boxes are in exactly the same position, so the circles overlap).

**Comment by patrickhlauke** [comment]
oh, hang on, completely missed that the two SVGs go to different bits...I somehow had it in my head that they're two disconnected bits of the same link. let me re-look at this with a clearer head...

**Comment by dd8** [comment]
Interestingly, it looks like you can have two identical looking SVGs, one of which passes while the other fails.

````
<!-- blue frame is a single undersized target - the blue frame bounding box is centred exactly on the red box -->
<svg width="500" height="500" viewBox="0 0 500 500"> 
    <a href="one"> <rect was-stroke-alignment="inner" stroke-width="14px" stroke="blue" fill="none" x="10" y="10" width="460" height="460"></rect> </a> 
    <a href="two"> <rect fill="red" x="240" y="240" width="20" height="20"></rect> </a> </svg>
```

```
<!-- blue frame is 4 different undersized targets - the blue bounding boxes are at the edge of the SVG and don't overlap the red box -->
    <svg width="500" height="500" viewBox="0 0 500 500"> 
        <a href="one-top"> <rect stroke-width="10px" stroke="blue" fill="none" x="10" y="10" width="460" height="5"></rect> </a> 
        <a href="one-left"> <rect stroke-width="10px" stroke="blue" fill="none" x="10" y="10" width="5" height="460"></rect> </a> 
        <a href="one-right"> <rect stroke-width="10px" stroke="blue" fill="none" x="470" y="10" width="5" height="460"></rect> </a> 
        <a href="one-bottom"> <rect stroke-width="10px" stroke="blue" fill="none" x="10" y="470" width="465" height="5"></rect> </a> 

        <a href="two"> <rect fill="red" x="240" y="240" width="20" height="20"></rect> </a> </svg>
```

**Comment by patrickhlauke** [comment]
the crux of the problem is, i think, the way the SC was simplified by using the bounding box, and not accounting for nested (but not overlapping/clipping, because of the "hole" in the middle) targets.

one possible way forward that would not need a complicated rewrite of the SC may be to expand the "clipping" example/case (figure 9 in https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html) to basically cover the more general "if one target is inside the *bounding box* of another target"

**Comment by dd8** [comment]
The same problem also happens with U shapes (take away one side) and L shapes (take away two sides) because the outer bounding box stays the same size.

A real world version of the first example I posted is a video control with a play button in the centre, and a thin strip around the video edge which displays it full screen.

BTW We're writing some test cases for ACT Rules, so more issues may crop up. Maybe worth holding off any rewrites until we've done a few more

**Comment by dd8** [comment]
The same problem also happens with U shapes (take away one side) and L shapes (take away two sides) because the outer bounding box stays the same size.

A real world version of the first example I posted is a video control with a play button in the centre, and a thin strip around the video edge which displays it full screen.

BTW We're writing some test cases for ACT Rules, so more issues may crop up. Maybe worth holding off any rewrites until we've done a few more

**Comment by patrickhlauke** [comment]
> The same problem also happens with U shapes (take away one side) and L shapes (take away two sides) because the outer bounding box stays the same size.

the core problem is that the simplification that was consciously made, using the bounding box, which avoided having to do the mind-bending actual geometric calculations for arbitrarily shaped targets that aren't rectangular, will not be accurate in exactly those situations that were avoided.

**Comment by dd8** [comment]
Another tricky example - this SVG looks like a % sign - with 2 `<a>` elements - one for the red slash and another for the blue circles either side of the slash.

The user agent has two event targets here - should the SC evaluate this as 2 targets or 3 targets? (i.e. should the blue circles be evaluated as one target or two, which affects pass/fail) 


```
        <svg width="50px" height="50px" viewBox="0 0 394.4 394.4">
            <a href="one">
                <path fill="red" d="M37.4,377.4c-5.223,0-10.438-1.992-14.423-5.977c-7.97-7.963-7.97-20.883,0-28.846l319.6-319.601
                    c7.97-7.97,20.876-7.97,28.846,0c7.97,7.962,7.97,20.882,0,28.845l-319.6,319.601C47.838,375.408,42.623,377.4,37.4,377.4z
                        "/>
            </a>

            <a href="two">
            <path fill="blue" d="M394.4,299.199c0-52.496-42.704-95.199-95.2-95.199S204,246.703,204,299.199s42.704,95.201,95.2,95.201
            S394.4,351.695,394.4,299.199z M353.601,299.199c0,29.996-24.405,54.4-54.4,54.4s-54.4-24.404-54.4-54.4
            c0-29.994,24.405-54.398,54.4-54.398S353.601,269.205,353.601,299.199z M190.4,95.2C190.4,42.704,147.696,0,95.2,0S0,42.704,0,95.2
            s42.704,95.2,95.2,95.2S190.4,147.696,190.4,95.2z M149.6,95.2c0,29.995-24.405,54.4-54.4,54.4s-54.4-24.405-54.4-54.4
            s24.405-54.4,54.4-54.4S149.6,65.206,149.6,95.2z"/>
            </a>
        </svg>
```

**Comment by patrickhlauke** [comment]
> Another tricky example - this SVG looks like a % sign - with 2 <a> elements - one for the red slash and another for the blue circles either side of the slash.

yeah, that's the other part that the SC/understanding don't quite make clear ... what to do with targets that are effectively split into multiple bits (regardless of how they're coded under the hood). my take there (not backed up by any wording in the documentation) is to ignore how things are actually coded in HTML/SVG/etc and (naively) look at the effect: that there's two separate circles that are two separate targets going to the same place, and evaluate them separately.

**Comment by dd8** [comment]
There are probably pros and cons to both approaches:

- splitting the blue circles into separate targets splits the bounding box, and produces two target circle which can overlap with nearby controls (maybe that's good in this scenario, and since both are "equivalent" this only triggers failures when both are close to something else)
- splitting them also means multiple click targets map to a single focusable target (which may have unintended side effects elsewhere - especially if the "target" definition is used in other SCs)
- this particular example gets tricker if the paths making up the % are turned into a single '%' character (you probably don't want to start treating parts of glyphs as targets, but then you get into debates about why the circles that look like they're part of a % glyph are treated as separate targets when they're drawn as a path but not when they're part of a % character)

**Comment by JAWS-test** [comment]
@dd8 If I remember correctly, we discussed all these issues before WCAG 2.2 was adopted. In the end, the WG refused to go into all the possible borderline cases (including those that I submitted). That's why we now have to live with the fact that, for example, 2 targets whose distance is large enough but whose center of the bounding box matches violate the SC.

I only see a need for clarification on one point, your last example, where it is a question of whether a target which does not form a contiguous area should count as 1 or 2 targets. I tend to think that it is 2 targets, but I also understand your counter-argument (last point in your list). I can't think of a good solution. In any case, the WG should decide this problem and add it to the Understanding! However, there may already be a decision on this, namely in the definition of target:

> If two or more targets are overlapping, the overlapping area should not be included in the measurement of the target size, except when the overlapping targets perform the same action or open the same page. 

If I understand correctly, this means that the two circles of your percentage sign count as one target

**Comment by dd8** [comment]
If the multiple targets interpretation works, then one way of resolving concave shapes is to treat them as if they're made up of simpler targets. So 

```
<svg width="500" height="500" viewBox="0 0 500 500"> 
    <a href="one"> <rect stroke-width="14px" stroke="blue" fill="none" x="10" y="10" width="460" height="460"></rect></a> 
    <a href="two"> <rect fill="red" x="240" y="240" width="20" height="20"></rect> </a> 
</svg>
```

is evaluated as if it was coded as:

```
<svg width="500" height="500" viewBox="0 0 500 500"> 
<!-- looks identical to above, but frame split into 4 sides -->
        <a href="one"> <rect stroke-width="10px" stroke="blue" fill="none" x="10" y="10" width="460" height="5"></rect> </a> 
        <a href="one"> <rect stroke-width="10px" stroke="blue" fill="none" x="10" y="10" width="5" height="460"></rect> </a> 
        <a href="one"> <rect stroke-width="10px" stroke="blue" fill="none" x="470" y="10" width="5" height="460"></rect> </a> 
        <a href="one"> <rect stroke-width="10px" stroke="blue" fill="none" x="10" y="470" width="465" height="5"></rect> </a> 

        <a href="two"> <rect fill="red" x="240" y="240" width="20" height="20"></rect> </a> 
</svg>
```

That doesn't seem a huge stretch since targets with identical appearance should be evaluated identically.

**Comment by JAWS-test** [comment]
> If the multiple targets interpretation works, then one way of resolving the issues with concave shaped targets is to treat them as if they're made up of simpler shapes

I don't think that would work, because the division would be arbitrary. With your square, it is obvious to divide it into four lines, but what if we really have an irregular shape (e.g. rivers on a map) or a circle - is it halved, quartered, eighth...?

**Comment by JAWS-test** [comment]
The only sensible solution I can think of is that targets are those that are located in an element (e.g. a link). As soon as I have 2 elements (e.g. 2 links), they are two separate targets, even if they look completely identical and have the same link target

**Comment by dd8** [comment]
> The only sensible solution I can think of is that targets are those that are located in an element (e.g. a link). As soon as I have 2 elements (e.g. 2 links), they are two separate targets, even if they look completely identical and have the same link target

This matches up with how event handling works in the browser. EventTarget objects that handles pointer events are attached to elements:
https://developer.mozilla.org/en-US/docs/Web/API/EventTarget

This also aligns with how keyboard focus works - which may turn out to be important.

That means there are two EventTargets in the example from https://github.com/w3c/wcag/issues/3615#issuecomment-1862476967

```
        <svg width="50px" height="50px" viewBox="0 0 394.4 394.4">
        <!-- the a element is an EventTarget for click events: painted areas inside the element are clickable -->
            <a href="one">
                <!-- red slash-->
                <path fill="red" d="M37.4,377.4c-5.223,0-10.438-1.992-14.423-5.977c-7.97-7.963-7.97-20.883,0-28.846l319.6-319.601
                    c7.97-7.97,20.876-7.97,28.846,0c7.97,7.962,7.97,20.882,0,28.845l-319.6,319.601C47.838,375.408,42.623,377.4,37.4,377.4z
                        "/>
            </a>

        <!-- the a element is an EventTarget for click events: painted areas inside the element are clickable -->
           <a href="two">
            <!-- two blue circles -->
            <path fill="blue" d="M394.4,299.199c0-52.496-42.704-95.199-95.2-95.199S204,246.703,204,299.199s42.704,95.201,95.2,95.201
            S394.4,351.695,394.4,299.199z M353.601,299.199c0,29.996-24.405,54.4-54.4,54.4s-54.4-24.404-54.4-54.4
            c0-29.994,24.405-54.398,54.4-54.398S353.601,269.205,353.601,299.199z M190.4,95.2C190.4,42.704,147.696,0,95.2,0S0,42.704,0,95.2
            s42.704,95.2,95.2,95.2S190.4,147.696,190.4,95.2z M149.6,95.2c0,29.995-24.405,54.4-54.4,54.4s-54.4-24.405-54.4-54.4
            s24.405-54.4,54.4-54.4S149.6,65.206,149.6,95.2z"/>
            </a>
        </svg>
```

For manual evaluation the target can identified by using DevTools Inspect - it highlights the bounding box (and shows two overlapping bounding boxes in the example above)

**Comment by aardrian** [comment]
I am looking at a donut chart with a nested donut chart. While not as clear-cut as the thing nested within a square, this is also not a hypothetical.

When some values in the donut chart are sufficiently big to describe an obtuse angle in the circle (though the design could easily make it impact smaller circle segments), the center of the bounding box intersects with the nested chart.

Landed on this issue because I was searching to see if there was guidance on this scenario. Some clarity would be appreciated.
